File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void inorder(TreeNode root,List<Integer>list){
        if(root==null){
            return;
        }
        inorder(root.left,list);
        list.add(root.val);
        inorder(root.right,list);
    }
    public boolean findTarget(TreeNode root, int k) {
        List<Integer>list=new ArrayList<>();
        inorder(root,list);

        int l=0;
        int r=list.size()-1;
        while(l<r){
            if(list.get(l)+list.get(r)==k){
                return true;
            }else if(list.get(l)+list.get(r)<k){
                l++;
            }else{
                r--;
            }
        }
        return false;
    }
}

File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_set<int>s;
    bool helper(TreeNode* root, int k){
        if(root==NULL){
            return false;
        }
        if(s.count(k-root->val)){
            return true;
        }
        s.insert(root->val);
        return helper(root->left,k)||helper(root->right,k);
    }
    bool findTarget(TreeNode* root, int k) {
        return helper(root,k);
    }
};